future-apps/staking/test/staking.js

```diff
@@ -0,0 +1,480 @@
+      const initialTimestamp = (await getBlock(initialBlockNumber)).timestamp
+      await app.lock(amount / 4, TIME_UNIT_SECONDS, initialTimestamp + time, other, '', '')
+      await app.lock(amount / 4, TIME_UNIT_SECONDS, initialTimestamp + time, other, '', '')
+
+      // unlock
+      await app.unlockAll(owner, { from: other })
+
+      assert.equal((await app.unlockedBalanceOf(owner)).valueOf(), amount, "Unlocked balance should match")
+      assert.equal((await app.locksCount(owner)).valueOf(), 0, "there shouldn't be locks")
+    })
+
+    it('unlocks all with no previous locks', async () => {
+      await app.unlockAll(owner, { from: other })
+      assert.equal((await app.locksCount(owner)).valueOf(), 0, "there shouldn't be locks")
+    })
+
+    it('tries to unlockAll but it only unlocks one', async () => {
+      const amount = 100
+      const time = 1000
+      await token.approve(app.address, amount)
+      await app.stake(amount, '')
+      await app.lockIndefinitely(amount / 2, other, '', '')
+      const initialBlockNumber = await getBlockNumber()
+      const initialTimestamp = (await getBlock(initialBlockNumber)).timestamp
+      await app.lock(amount / 4, TIME_UNIT_SECONDS, initialTimestamp + time, other, '', '')
+
+      await timeTravel(time + 1)
+      // unlock
+      await app.unlockAll(owner)
+
+      assert.equal((await app.unlockedBalanceOf(owner)).valueOf(), amount / 2, "Unlocked balance should match")
+      assert.equal((await app.locksCount(owner)).valueOf(), 1, "there shouldn't be locks")
+    })
+
+    it('fails trying to unlockAllOrNone', async () => {
+      const amount = 100
+      const time = 1000
+      await token.approve(app.address, amount)
+      await app.stake(amount, '')
+      await app.lockIndefinitely(amount / 2, other, '', '')
+      const initialBlockNumber = await getBlockNumber()
+      const initialTimestamp = (await getBlock(initialBlockNumber)).timestamp
+      await app.lock(amount / 4, TIME_UNIT_SECONDS, initialTimestamp + time, other, '', '')
+
+      await timeTravel(time + 1)
+      // unlock
+      return assertRevert(async () => {
+        await app.unlockAllOrNone(owner)
+      })
+    })
+
+    it('moves tokens', async () => {
+      const amount = 100
+      const time = 1000
+      await token.approve(app.address, amount)
+      await app.stake(amount, '')
+      await app.moveTokens(owner, other, amount / 2)
+    })
+
+    it('fails moving 0 tokens', async () => {
+      const amount = 100
+      await token.approve(app.address, amount)
+      await app.stake(amount, '')
+      return assertRevert(async () => {
+        await app.moveTokens(owner, other, 0)
+      })
+    })
+
+    it('fails moving more tokens than staked', async () => {
+      const amount = 100
+      await token.approve(app.address, amount)
+      await app.stake(amount, '')
+      return assertRevert(async () => {
+        await app.moveTokens(owner, other, amount + 1)
+      })
+    })
```

Comment:
Add a test for failing to move more tokens from an account than unlocked.

Reply:
Done.